tests/test_server.py

```diff
@@ -25,22 +19,49 @@ def negotiate_and_ntlm_auth():
     return get_auth_response('NTLM', advertise_nego_and_ntlm=True)
 
 def get_auth_response(auth_type, advertise_nego_and_ntlm=False):
-    response_headers = {'WWW-Authenticate':auth_type if not advertise_nego_and_ntlm else 'Negotiate, NTLM'}
+    # Set the default responses
+    response_headers = {'WWW-Authenticate': auth_type if not advertise_nego_and_ntlm else 'Negotiate, NTLM'}
     status_code = 401
     response = "auth with '%s\\%s':'%s'" % (domain, username, password)
 
-    # 2nd request
-    if request.headers.get('Authorization','') == REQUEST_2_TEMPLATE.format(auth_type):
-        response_headers = {'WWW-Authenticate':RESPONSE_2_TEMPLATE.format(auth_type)}
+    # Get the actual header that is returned by requests_ntlm
+    actual_header = request.headers.get('Authorization', '')
+
+    # Check what the message type is from the header
+    if actual_header == '':
+        # This is the initial connection, need to return a 401
+        message_type = None
+    else:
+        message_type = get_message_type(auth_type, actual_header)
+
+    # Validate that the message type is either a negotiate or authenticate message and act accordingly
+    if message_type == 1:
+        # Add the challenge token and return it with the response headers
+        challenge_response = ('TlRMTVNTUAACAAAAAwAMADgAAAAzgoriASNFZ4mrze8AAAA'
+                              'AAAAAACQAJABEAAAABgBwFwAAAA9TAGUAcgB2AGUAcgACAA'
+                              'wARABvAG0AYQBpAG4AAQAMAFMAZQByAHYAZQByAAAAAAA=')
+        challenge_header = auth_type + ' ' + challenge_response
+        response_headers = {'WWW-Authenticate': challenge_header}
         status_code = 401
 
-    # 3rd request
-    elif request.headers.get('Authorization','') == REQUEST_3_TEMPLATE.format(auth_type):
+    # Server received the authenticate token and validates it matches what we expect
+    elif message_type == 3:
         response_headers = {}
         status_code = 200
-        response = "authed"
+        response = 'authed'
```

Comment:
Isn't that what this section is for, we always return 401 unless the message is a negotiate message or an authenticate message? Are you suggesting we have some sort of counter where we expect the first to be nothing, return 401, expect auth, return challenge, expect negotiate, return success?